File: leetcode/javaStuff/middleOfLinkedList/MiddleLinkedList.java
import java.util.*;

public class MiddleLinkedList {
  public static void main(String[] args) {
    ListNode a = new ListNode(1);
    ListNode b = new ListNode(2);
    ListNode c = new ListNode(3);
    ListNode d = new ListNode(4);
    ListNode e = new ListNode(5);

    a.next = b;
    b.next = c;
    c.next = d;
    d.next = e;

    ArrayList<ListNode> test = getFullList(a);
    System.out.println(test.size());
  }

  // public static ListNode findMiddleNode(ListNode head) {
  //   ArrayList<ListNode> allNodes = getFullList(head);
  //
  //
  // }

  public static ArrayList<ListNode> getFullList(ListNode head) {
    ArrayList<ListNode> answer = new ArrayList<ListNode>();
    ListNode currentNode = head;

    while(currentNode != null) {
      answer.add(currentNode);
      currentNode = currentNode.next;
    }

    return answer;
  }

}

File: leetcode/middleLinkedList.js
// O(n) time and O(1) space

class ListNode {
  constructor(val) {
    this.val = val;
    this.next = undefined;
  }
}

function middleLinkedList(head) {
  let slowPointer = head;
  let fastPointer = head;

  while(fastPointer.next !== undefined) {
    slowPointer = slowPointer.next;
    fastPointer = fastPointer.next.next;
  }

  return slowPointer;
}


let a = new ListNode(1);
let b = new ListNode(2);
let c = new ListNode(3);
let d = new ListNode(4);
let e = new ListNode(5);

a.next = b;
b.next = c;
c.next = d;
d.next = e;

console.log(middleLinkedList(a));
